src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,538 @@
+            LOGGER.info("{} has container ID {}, but the container does not exist in any docker cloud. Will remove node.", dtn, dtn.getContainerId());
+
+            try {
+                removeNode(dtn);
+                executionStatistics.nodesRemoved++;
+            } catch (IOException e) {
+                LOGGER.warn(String.format("Failed to remove orphaned %s.", dtn), e);
+                executionStatistics.nodesRemovedFailed++;
+            }
+        }
+    }
+
+    private void checkForTimeout(Instant startedTimestamp) {
+        final Instant now = clock.instant();
+        Duration runtime = Duration.between(startedTimestamp, now);
+        
+        if (runtime.compareTo(PROCESSING_TIMEOUT_IN_MS) > 0) {
+            throw new WatchdogProcessingTimeout();
+        }
+    }
+    
+    private static class Statistics {
+        public long executions;
+        public long containersRemovedGracefully;
+        public long containersRemovedGracefullyRuntimeSum;
+        public long containersRemovedForce;
+        public long containersRemovedForceRuntimeSum;
+        public long containersRemovedFailed;
+        public long nodesRemoved;
+        public long nodesRemovedFailed;
+        public long processingTimeout;
+        public long overallRuntime;
+    }
+    
+    private void logStatistics() {
+        LOGGER.info("Watchdog Statistics: "
+                + "Number of overall executions: {}, "
+                + "Executions with processing timeout: {}, "
+                + "Containers removed gracefully: {}, "
+                + "Containers removed with force: {}, "
+                + "Containers removal failed: {}, "
+                + "Nodes removed successfully: {}, "
+                + "Nodes removal failed: {}, "
+                + "Container removal average duration (gracefully): {} ms, "
+                + "Container removal average duration (force): {} ms, "
+                + "Average overall runtime of watchdog: {} ms",
+                executionStatistics.executions, 
+                executionStatistics.processingTimeout,
+                executionStatistics.containersRemovedGracefully,
+                executionStatistics.containersRemovedForce,
+                executionStatistics.containersRemovedFailed,
+                executionStatistics.nodesRemoved,
+                executionStatistics.nodesRemovedFailed,
+                executionStatistics.containersRemovedGracefully != 0 ? new Long(executionStatistics.containersRemovedGracefullyRuntimeSum / executionStatistics.containersRemovedGracefully).toString() : 0,
+                executionStatistics.containersRemovedForce != 0 ? new Long(executionStatistics.containersRemovedForceRuntimeSum / executionStatistics.containersRemovedForce).toString() : 0,
+                executionStatistics.executions != 0 ? new Long(executionStatistics.overallRuntime / executionStatistics.executions).toString() : 0
+                );
+    }
```

Comment:
This is the sort of logic that _may_ be better placed inside the Statistics class than outside it.

Reply:
see also 9af913b48aa2693583334a367f3e26f3c7643fee (and my comment above)